### DISCUSSION TITLE
Testing data cached by useQuery.

### DISCUSSION THREAD

**Discussion by abelta** [TanStack/query#1883] [Q&A]
I'm going crazy finding a solution for this one even though I would guess it's a pretty common use.
I have a custom react-query hook that accesses the currently logged in user in my application like this:

```javascript
// useCurrentUser.js
import { useQuery } from 'react-query'

export default () => {
  const { data } = useQuery('login')
  return data
}
```

And then I try to test it mocking the entire module react-query first and then the implementation of useQuery so I can control what it returns (can't think of a more standard way than doing just that).

```javascript
//useCurrentUser.test.js
import { renderHook } from '@testing-library/react-hooks'
import { useQuery } from 'react-query'
import { useCurrentUser } from 'hooks'

jest.mock('react-query', () => ({
  useQuery: () => jest.fn(),
}))

describe('useCurrentUser', () => {

  it('returns current user', () => {
    const user = { id: 1, name: 'Ana' }
    useQuery.mockImplementation(() => user)
    const { result } = renderHook(() => useCurrentUser())
    expect(result).toBe(user)
  })
})
```

However it seems like this is not the right way to mock an NPM module because I get an error message: **_reactQuery.useQuery.mockImplementation is not a function**.

I am asking because it seems to me like this should be a pretty basic functionality that most users would like to test but is is not covered in the examples in the documentation. Any ideas?

**Comment by TkDodo** [comment]
I think you should just mock the actual network request or fetch-function that is happening, not the whole react-query module.

Here is how you can test a custom hook with react-query:

```
import * as React from "react";
import { QueryClient, QueryClientProvider, useQuery } from "react-query";
import { renderHook } from "@testing-library/react-hooks";
export function useCustomHook() {
  return useQuery("customHook", () => "Hello");
}

const createWrapper = () => {
  const queryClient = new QueryClient();
  const wrapper = ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return wrapper;
};

test("should work", async () => {
  const { result, waitFor } = renderHook(() => useCustomHook(), {
    wrapper: createWrapper()
  });

  await waitFor(() => result.current.isSuccess);

  expect(result.current.data).toEqual("Hello");
});
```

**Reply by TkDodo** [reply]
Here is a more comprehensive example that mocks the api layer with `jest.spyOn`

https://codesandbox.io/s/react-query-test-with-mocking-3qoy9?file=/src/app.test.tsx

**Reply by abelta** [reply]
Thank you for the answer, @TkDodo  but I'm not sure if this is the right way to go when the hook doesn't actually make any calls and all it does is read the cache stored from other calls, is it?

**Reply by TkDodo** [reply]
I don't quite follow - how does your custom hook look if uses `useQuery` but doesn't make a request ?

**Reply by abelta** [reply]
A different hook (login) does the call and then the returned value (current user) is stored in the cache.
I get it it is one of the encouraged practices with react-query to use it as a cache.

**Reply by TkDodo** [reply]
okay so you have a mutation (login), that mutation fills the cache, and then your query reads that value from the cache?

you can also fill the cache for testing purpose via `queryClient.setQueryData`, similar to what you're probably doing in the mutation.

**Reply by abelta** [reply]
I am sorry if I'm not explaining correctly: **useLogin** is a mutator (not included in the example). It posts for a login and gets as a result the data of the authenticated user that then it stores in the react-query cache in the key "login". Then **useCurrentUser** all it does it access this key in the cache but it never makes any calls.

**Reply by abelta** [reply]
Just to make it clear one more time, even if the interaction between the hooks is not understood: in **useCurrentUser** I'm accessing the cache and I want to test for that.

**Reply by TkDodo** [reply]
how do you access the cache? even if you never intend to make a request, I think you need a `useQuery` that sets up the subscription. queryClient.getQueryData will not update your component if data in the cache changes.

If you never intend to make a request _at all_ for that data, then don't write the user data to the query cache - put it into react context or another local state manager. The query cache is really just a caching mechanism for data retrieved by the server via `useQuery`...

**Reply by abelta** [reply]
The system I have set up actually does work, it testing it (barely a three line hook) that I can't find any example of.

```javascript
// useCurrentUser.js
import { useQuery } from 'react-query'

export default () => {
  const { data } = useQuery('login')
  return data
}
```

**Reply by TkDodo** [reply]
do you have a default query function setup? because per default, this will trigger background refetches when the component mounts, on window re-focus etc ...

if you don't have an actual async function that makes a request, you are kinda abusing react-query for synchronous state (just reading from a js Map basically). Sure, it works, but it's not what react-query is intended for :)

**Reply by abelta** [reply]
I'm willing to admit it's a bad use of the library but still it would be cool to be able to test cache access.

**Reply by TkDodo** [reply]
okay, fair enough :)

I would say that this should still work:

> you can also fill the cache for testing purpose via queryClient.setQueryData, similar to what you're probably doing in the mutation.

I'd do that in `createWrapper`:

```diff
const createWrapper = () => {
  const queryClient = new QueryClient();
+  queryClient.setQueryData('login', mockLoginObject)
  const wrapper = ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return wrapper;
};

**Reply by abelta** [reply]
Finally managed to test it while maintaining my initial structure just by accounting for the format useQuery uses when returning a result (inside a "data" key).
This is what I was looking for:

```javascript
import { renderHook } from '@testing-library/react-hooks'
import { useCurrentUser } from 'hooks'

jest.mock('react-query', () => ({
  useQuery: () => ({ data: { id: 1, name: 'Ana' } }),
}))

describe('useCurrentUser', () => {
  describe('user is logged in', () => {
    it('returns current user', () => {
      const { result } = renderHook(() => useCurrentUser())
      expect(result.current).toEqual({ id: 1, name: 'Ana' })
    })
  })
})

```
